### ISSUE TITLE
BC-breaking change to symint range constraints from 2.7 -> 2.8

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#157921]
```py
import torch

embed = torch.randn(2, 8192)
x = torch.zeros(8192)

torch._dynamo.mark_dynamic(x, 0)

@torch.compile(backend="aot_eager", fullgraph=True)
def f(embedding_indices, x):
    added_tokens_mask = torch.where(x > 10000, 1, 0)
    ei = torch.narrow(embedding_indices, 1, 0, x.size(0))
    return ei.clone()

f(embed, x)
```
This errors in 2.8 but passes in 2.7..

cc @chauhang @penguinwu @ezyang @bobrenjc93

**Comment by zou3519** [comment]
Adding to milestone for investigation, this might just be that we need to document the change in our release notes.

**Comment by zou3519** [comment]
I think this is causing some vLLM test failures on 2.8, e.g. `tests/lora/test_quant_model.py::test_quant_model_lora[model0]`, so we likely need to do something (in code) about it.

The interesting thing is that this specialization doesn't actually matter in the context of vLLM, but other specializations do..

**Comment by ezyang** [comment]
can you paste the error

**Comment by zou3519** [comment]
Here's the error on 2.8rc: https://gist.github.com/zou3519/48513baac3976a966efa5d7c6ab21bd8

Context:
- We investigated a vLLM bug internally. This repro came out of this thread [Meta-internal thread](https://fb.workplace.com/groups/1240828337440358/permalink/1266866794836512/). The internal vLLM bug might be the same thing as the vLLM test failure I reported above.
- Laith diagnosed the internal bug to contiguous checks in clone
- I tried out the repro on 2.7, and noticed that it didn't fail.

**Comment by zou3519** [comment]
I have bisected this to https://github.com/pytorch/pytorch/pull/152661, thoughts @aorenste ?

**Comment by aorenste** [comment]
> I have bisected this to [#152661](https://github.com/pytorch/pytorch/pull/152661), thoughts [@aorenste](https://github.com/aorenste) ?

I'll take a look.

**Comment by laithsakka** [comment]
there is 4 specializations points on the shape,
https://github.com/pytorch/pytorch/pull/157965

probably best way to handle this is unabcked or backed_size_oblivious for lora. 

chatted with @zou3519  and we agreed we shall not revert https://github.com/pytorch/pytorch/pull/152661

**Comment by aorenste** [comment]
@zou3519 @laithsakka 
Although the issue bisects to #152661 it actually is more complicated than that: 
1. #152661 just fixes the cache - the cache key was incorrect before. If I back out #152661 and then remove the cache entirely then the problem still happens (so it's an existing problem that the broken cache was masking)
2. If I back out #152661 on main then the problem still happens because of something that changed in #153432. So backing it out is not easily feasable at this point anyway because of other changes after it.
3. Because #153432 isn't easily backed (lots of conflicts) it I couldn't tell if backing them both out was possible.

It may be worth checking out just before #152661, disabling the cache and trying to fix the error to see what the underlying issue is.

**Comment by atalman** [comment]
HI @aorenste @zou3519 and @malfet According to First point in https://github.com/pytorch/pytorch/issues/157921#issuecomment-3055231416 this is pre-existing issue in 2.7 Should we de-milestone it from release 2.8 ?

**Comment by malfet** [comment]
@atalman Not sure what do you mean, I can clearly see that snippet above compiles fine in 2.7.1 and fails in 2.8.0

**Comment by aorenste** [comment]
The problem here is that as far as I can tell this repro SHOULD fail:
It's using `mark_dynamic(x, 0)` which says that it wants `x.size(0)` to be dynamic. But then it specializes on `x.size(0)` (because it's a parameter to `torch.narrow`). The definition of `mark_dynamic()` is that if you specialize to a single value it will fail. If you're okay with it being specialized then you're supposed to use `maybe_mark_dynamic()`.

So 2.7 was definitely the incorrect behavior for this repro and we're now doing the documented correct behavior.

**Comment by malfet** [comment]
@aorenste can you point to a line of documentation that says it's an invalid behavior?
The public document I can find is https://docs.pytorch.org/docs/stable/torch.compiler_dynamic_shapes.html, which says
>  If you know ahead of time something will be dynamic, you can skip the first recompile with `torch._dynamo.mark_dynamic(tensor, dim)`. If you know ahead of time the min and max value this dimension can take, you can specify `torch._dynamo.mark_dynamic(tensor, dim, min=min, max=max)`

I.e. user tells that x shape is dynamic, and I'm not sure why 2.8 compiler says that code above specializes it to be constant...

**Comment by malfet** [comment]
Hmm, marking dimension of emebed as dynamic makes code compilable on 2.7 and 2.8 and actually produces the constrains I've expected to have when only one dim were marked as dynamic
```python
import torch

embed = torch.randn(2, 8192)
x = torch.zeros(8192)

torch._dynamo.mark_dynamic(x, 0)
torch._dynamo.mark_dynamic(embed, 1)

@torch.compile(backend="aot_eager", fullgraph=True)
def f(embedding_indices, x):
    added_tokens_mask = torch.where(x > 10000, 1, 0)
    ei = torch.narrow(embedding_indices, 1, 0, x.size(0))
    return ei.clone()

f(embed, x)
```

@zou3519 should this just be changed in the vLLM code? But PR that introduces the check should be marked as BC-breaking (imo)

**Comment by aorenste** [comment]
@malfet I guess I don't see it in the docs - but it's definitely expected behavior - for example: https://github.com/pytorch/pytorch/issues/102814

The error message itself says what to do:
```
  - You marked L['x'].size()[0] as dynamic but your code specialized it to be a
    constant (8192). If you're using mark_dynamic, either remove it or use
    maybe_mark_dynamic. If you're using Dim.DYNAMIC, replace it with either
    Dim.STATIC or Dim.AUTO.
```

> user tells that x shape is dynamic, and I'm not sure why 2.8 compiler says that code above specializes it to be constant

I believe using `x.size(0)` as a parameter to `torch.narrow` is what is specializing it. Rather than something like `x > 10000` where it's a  range, using it as a parameter makes it an exact value (I'm guessing - @laithsakka would know for sure).

But the change in behavior between 2.7 and 2.8 here isn't in the dynamic behavior - that didn't change. The problem is that in 2.7 the cache for evaluating dynamic shapes doesn't include the "suppress guards" flag in the key even though it uses it when computing the value - and in some cases (in this particular case) the code calls it twice - once w/ "suppress guards=True" as a test and then again w/ "suppress guards=False" when it is actually building the fx graph. The problem is that because the first call had "suppress guards=True" it never built the guards and the second call is just returning the cached value and never building the guards that the resulting graph is actually relying on. And the cache is just a 128 entry cache - so the behavior can be nondeterministic depending on how much stuff you did between the two checks.

**Comment by zou3519** [comment]
@malfet we updated vLLM to be compatible with this bug fix. There indeed was some code that worked in 2.7.1 but failed in 2.8.

Also, I would classify this as a "bug fix". We fixed something that was broken and there is no way to achieve the old behavior.

I think the action item here is to figure out if we want to mention something around this in the release notes. Do we include bug fixes in the BC-breaking section of the release notes? @albanD. If not, we want to check that this is in the bug fix section.

**Comment by albanD** [comment]
We should include in the bc-breaking section anything we expect is going to break legitimate usage.
As vLLM showed, this one is a legitimate usage. So we should definitely put it there.

**Comment by zou3519** [comment]
@laithsakka @aorenste could one of you add a BC-breaking note to the release notes please? Then we can close out this issue.   Should coordinate with @jbschlosser who is leading the release notes process (I don't know what stage we're in)

**Comment by aorenste** [comment]
> [@laithsakka](https://github.com/laithsakka) [@aorenste](https://github.com/aorenste) could one of you add a BC-breaking note to the release notes please? Then we can close out this issue. Should coordinate with [@jbschlosser](https://github.com/jbschlosser) who is leading the release notes process (I don't know what stage we're in)

https://github.com/pytorch-labs/torch-release-notes/pull/51
